### ISSUE TITLE
[feature-request] File/block-scoped specific noqa directives

### ISSUE THREAD

**Issue by smackesey** [charliermarsh/ruff#1054]
Ruff's README says this about [ignoring errors](https://github.com/charliermarsh/ruff#ignoring-errors):

>To ignore all errors across an entire file, Ruff supports Flake8's # flake8: noqa directive (or, equivalently, # ruff: noqa). Adding either of those directives to any part of a file will disable error reporting for the entire file.
>
>For targeted exclusions across entire files (e.g., "Ignore all F841 violations in /path/to/file.py"), see the [per-file-ignores](https://github.com/charliermarsh/ruff#per-file-ignores) configuration setting.

It is much easier to maintain "per-file-ignores" when they are _in_ the file they refer to, rather than a config file. Pylint lets you scope `# pylint: disable=<code>` (equivalent of `noqa`) to the file and [block level](https://pylint.pycqa.org/en/latest/user_guide/messages/message_control.html#block-disables):

```
### All the violations on a single line

a, b = ... # pylint: disable=unbalanced-tuple-unpacking

### All the violations on the following line

# pylint: disable-next=unbalanced-tuple-unpacking
a, b = ...

### All the violations in a single scope

def test():
    # Disable all the no-member violations in this function
    # pylint: disable=no-member
    ...

### All the violations in a block. For instance, each separate branch of an if statement is considered a separate block, as in the following example:

def meth5(self):
    # pylint: disable=no-member
    # no error
    print(self.bla)
    if self.blop:
        # pylint: enable=no-member
        # enable all no-members for this block
        print(self.blip)
```

Would be really nice for ruff to be able to do this with `# ruff: noqa <code>` comments. Particularly useful for:

- rules on `print` calls, which sometimes you want to allow in a particular file or function scope but generally want to prohibit throughout rest of codebase.
- temporarily disabling import sorting (you need this to replicate `# isort: off/on` comments

**Comment by charliermarsh** [comment]
(Just to briefly mention: we do support `# isort: off/on` comments. They should be respected as of now.)

**Comment by smackesey** [comment]
@charliermarsh Oh I didn't realize-- do you support all the isort [action comments](https://pycqa.github.io/isort/docs/configuration/action_comments.html), like `isort: split` and `isort: skip_file`?

**Comment by smackesey** [comment]
Hmm `# isort: off` doesn't seem to be working. Ruff `0.0.160` reports import errors on the below file-- shouldn't the `# isort: off` at the top disable any `I001` error for the whole file?

```
# isort: off
# pylint: disable=import-error

from dagster import (
    load_assets_from_package_module,
)

# start_example

import my_package.cereal as cereal

cereal_assets = load_assets_from_package_module(
    cereal,
    group_name="cereal_assets",
)

# end_example
```

**Comment by charliermarsh** [comment]
It should work if you add `# isort: on` to the bottom (which should probably be considered a bug).

**Comment by smackesey** [comment]
> I could add support for split and/or skip_file if those are impactful for you.

We do use `split`. IIUC `skip_file` is the same as putting `off` at the top. I do think if you are going to support any of them, you should support them all, since they're not `ruff`-specific error codes but `isort`-specific commands. Would certainly help people port from `isort`.

**Comment by smackesey** [comment]
>It should work if you add # isort: on to the bottom (which should probably be considered a bug).

Definitiely a bug going by the `isort` docs for `off`:

>Turns isort parsing off. Every line after an # isort: off statement will be passed along unchanged until an # isort: on comment or the end of the file.

**Comment by charliermarsh** [comment]
Yeah. Can probably fix and add the rest of these today.

**Comment by charliermarsh** [comment]
@smackesey - What if we support [ESLint-like on/off comments](https://eslint.org/docs/latest/user-guide/configuring/rules#disabling-rules)? I'd prefer to do something that's distinct from language semantics (like block-scoped rules) if possible -- or, at least, it's easier, and I think it could be more intuitive too.

**Comment by charliermarsh** [comment]
I'd like to support this soon. I'm still leaning towards doing on-off comments that are based purely on lines and not scope, for simplicity.

**Comment by smackesey** [comment]
Is this on the radar for anytime soon? BTW `ruff` has been fantastic for us since we adopted it at Dagster.

**Comment by charliermarsh** [comment]
I haven't been thinking about it lately :joy: but I could re-prioritize it! I think the implementation itself isn't too hard, it's mostly coming up with a clean API to support going forward.

**Comment by smackesey** [comment]
>coming up with a clean API to support going forward

Forget about block scoping for the moment (some variant of on/off comments would be good though)-- for file-scoping, why not just allow no `# noqa: <whatever>` at the top of the file to scope to the file?

**Comment by charliermarsh** [comment]
Yeah we can support that. Flake8 intentionally doesn't but I'm not certain why -- I think because they recommend you use `per-file-ignores` for that, but I don't think it's a huge burden for us to support it.

**Comment by charliermarsh** [comment]
I'm thinking we can do:

```py
# ruff: noqa: F401
```

...to disable F401 for the entire file.

We already support:

```py
# ruff: noqa
```

...to disable linting for the _entire_ file, so this feels like a reasonable blend of that syntax and `# noqa: F401` for inline ignores.

**Comment by smackesey** [comment]
Sounds good-- for multiple codes I'd recommend:

```
# ruff: noqa: <code1>,<code2>,...
```

**Comment by charliermarsh** [comment]
Yeah sounds good -- so similar to the inline form of `# noqa: <code1>, <code2>, ...`.

**Comment by charliermarsh** [comment]
Working on file-scoped directives now.
